src/picker_gui.cpp

```diff
@@ -0,0 +1,644 @@
+	 * PickerWindow::ConstructWindow must be called by the inheriting window. */
+}
+
+void PickerWindow::ConstructWindow()
+{
+	this->CreateNestedTree();
+
+	/* Test if pickers should be active.*/
+	bool isActive = this->callbacks.IsActive();
+
+	/* Functionality depends on widgets being present, not window class. */
+	this->has_class_picker = isActive && this->GetWidget<NWidgetBase>(WID_PW_CLASS_LIST) != nullptr && this->callbacks.HasClassChoice();
+	this->has_type_picker = isActive && this->GetWidget<NWidgetBase>(WID_PW_TYPE_MATRIX) != nullptr;
+
+	if (this->has_class_picker) {
+		this->GetWidget<NWidgetCore>(WID_PW_CLASS_LIST)->tool_tip = this->callbacks.GetClassTooltip();
+
+		this->querystrings[WID_PW_CLASS_FILTER] = &this->class_editbox;
+	} else {
+		if (auto *nwid = this->GetWidget<NWidgetStacked>(WID_PW_CLASS_SEL); nwid != nullptr) nwid->SetDisplayedPlane(SZSP_VERTICAL);
+	}
+
+	this->class_editbox.cancel_button = QueryString::ACTION_CLEAR;
+	this->class_string_filter.SetFilterTerm(this->class_editbox.text.buf);
+	this->class_string_filter.callbacks = &this->callbacks;
+
+	this->classes.SetListing(this->callbacks.class_last_sorting);
+	this->classes.SetFiltering(this->callbacks.class_last_filtering);
+	this->classes.SetSortFuncs(_class_sorter_funcs);
+	this->classes.SetFilterFuncs(_class_filter_funcs);
+
+	if (this->has_type_picker) {
+		/* Update used and saved type information. */
+		this->callbacks.saved = this->callbacks.UpdateSavedItems(this->callbacks.saved);
+		this->callbacks.used.clear();
+		this->callbacks.FillUsedItems(this->callbacks.used);
+
+		SetWidgetDisabledState(WID_PW_MODE_ALL, !this->callbacks.HasClassChoice());
+
+		this->GetWidget<NWidgetCore>(WID_PW_TYPE_ITEM)->tool_tip = this->callbacks.GetTypeTooltip();
+
+		auto *matrix = this->GetWidget<NWidgetMatrix>(WID_PW_TYPE_MATRIX);
+		matrix->SetScrollbar(this->GetScrollbar(WID_PW_TYPE_SCROLL));
+
+		this->querystrings[WID_PW_TYPE_FILTER] = &this->type_editbox;
+	} else {
+		if (auto *nwid = this->GetWidget<NWidgetStacked>(WID_PW_TYPE_SEL); nwid != nullptr) nwid->SetDisplayedPlane(SZSP_VERTICAL);
+	}
+
+	this->type_editbox.cancel_button = QueryString::ACTION_CLEAR;
+	this->type_string_filter.SetFilterTerm(this->type_editbox.text.buf);
+	this->type_string_filter.callbacks = &this->callbacks;
+
+	this->types.SetListing(this->callbacks.type_last_sorting);
+	this->types.SetFiltering(this->callbacks.type_last_filtering);
+	this->types.SetSortFuncs(_type_sorter_funcs);
+	this->types.SetFilterFuncs(_type_filter_funcs);
+
+	this->FinishInitNested(this->window_number);
+
+	this->InvalidateData(PFI_CLASS | PFI_TYPE | PFI_POSITION | PFI_VALIDATE);
+}
+
+void PickerWindow::Close([[maybe_unused]] int data)
+{
+	this->callbacks.Close(data);
+	this->PickerWindowBase::Close(data);
+}
+
+void PickerWindow::UpdateWidgetSize(WidgetID widget, Dimension &size, [[maybe_unused]] const Dimension &padding, [[maybe_unused]] Dimension &fill, [[maybe_unused]] Dimension &resize)
```

Comment:
`padding` and `resize` are used.
```suggestion
void PickerWindow::UpdateWidgetSize(WidgetID widget, Dimension &size, const Dimension &padding, [[maybe_unused]] Dimension &fill, Dimension &resize)
```